types/wlr_input_method_v2.c

```diff
@@ -101,10 +122,196 @@ static void im_get_input_popup_surface(struct wl_client *client,
+		void *data) {
+	struct wlr_input_method_keyboard_grab_v2 *keyboard_grab =
+		wl_container_of(listener, keyboard_grab, keymap_listener);
+	wlr_input_method_keyboard_grab_v2_set_keyboard(keyboard_grab, NULL);
+}
+
+void wlr_input_method_keyboard_grab_v2_set_keyboard(
+		struct wlr_input_method_keyboard_grab_v2 *keyboard_grab,
+		struct wlr_keyboard *keyboard) {
+	if (keyboard == keyboard_grab->keyboard) {
+		return;
+	}
+
+	if (keyboard_grab->keyboard) {
+		wl_list_remove(&keyboard_grab->keymap_listener.link);
+		wl_list_remove(&keyboard_grab->repeat_info_listener.link);
+		wl_list_remove(&keyboard_grab->destroy_listener.link);
+	}
+
+	if (keyboard) {
+		if (keyboard_grab->keyboard == NULL ||
+				strcmp(keyboard_grab->keyboard->keymap_string,
+				keyboard->keymap_string) != 0) {
+			// send keymap only if it is changed, or if input method is not
+			// aware that it did not change and blindly send it back with
+			// virtual keyboard, it may cause an infinite recursion.
+			if (!keyboard_grab_send_keymap(keyboard_grab, keyboard)) {
+				wlr_log(WLR_ERROR, "Failed to send keymap for input-method keyboard grab");
+				return;
+			}
+		}
+		keyboard_grab_send_repeat_info(keyboard_grab, keyboard);
+		keyboard_grab->keymap_listener.notify = handle_keyboard_keymap;
+		wl_signal_add(&keyboard->events.keymap,
+			&keyboard_grab->keymap_listener);
+		keyboard_grab->repeat_info_listener.notify =
+			handle_keyboard_repeat_info;
+		wl_signal_add(&keyboard->events.repeat_info,
+			&keyboard_grab->repeat_info_listener);
+		keyboard_grab->destroy_listener.notify =
+			handle_keyboard_destroy;
+	}
+
+	keyboard_grab->keyboard = keyboard;
+};
+
+void wlr_input_method_keyboard_grab_v2_destroy(
+		struct wlr_input_method_keyboard_grab_v2 *keyboard_grab) {
+	keyboard_grab_destroy(keyboard_grab);
+}
 static void im_grab_keyboard(struct wl_client *client,
 		struct wl_resource *resource, uint32_t keyboard) {
-	wlr_log(WLR_INFO, "Stub: zwp_input_method_v2::grab_keyboard");
+	struct wlr_input_method_v2 *input_method =
+		input_method_from_resource(resource);
+	if (!input_method) {
+		return;
+	}
+	if (input_method->keyboard_grab) {
+		// Already grabbed
+		return;
+	}
+	struct wlr_input_method_keyboard_grab_v2 *keyboard_grab =
+		calloc(1, sizeof(struct wlr_input_method_keyboard_grab_v2));
+	if (!keyboard_grab) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+	struct wl_resource *keyboard_grab_resource = wl_resource_create(
+		client, &zwp_input_method_keyboard_grab_v2_interface,
+		wl_resource_get_version(resource), keyboard);
+	if (keyboard_grab_resource == NULL) {
+		free(keyboard_grab);
+		wl_client_post_no_memory(client);
+		return;
+	}
+	wl_resource_set_implementation(keyboard_grab_resource,
+		&keyboard_grab_impl, keyboard_grab,
+		keyboard_grab_resource_destroy);
+	keyboard_grab->resource = keyboard_grab_resource;
+	keyboard_grab->input_method = input_method;
+	keyboard_grab->seat_client = wlr_seat_client_for_wl_client(
```

Comment:
What if the seat client is destroyed? We should setup a destroy listener.

Reply:
This is already handled by input method and will make it (and the grab, if exists) inert.

Reply:
The seat client may be destroyed before the seat, so I still think a destroy listener is necessary to avoid invalid pointers.

Reply:
The `seat_destroy` listener in `wlr_input_method_v2` actually listens to seat client destroy instead of the seat.

Reply:
Ah, okay. Can we change `wlr_input_method_v2` to have a `struct wlr_seat_client *seat_client` field instead of a `seat` field, and rename `seat_destroy` to `seat_client_destroy`?

Reply:
Gentle ping :)

Reply:
Done. Changes for sway side for seat_client instead of seat has also been added to sway side PR.